// @framer/jongheepaper.iphone-x-keyboard/code/Key.tsx
import * as React from "react"
import { Frame, addPropertyControls, ControlType } from "framer"
import { shiftKey, backKey } from "./Icons"
// Open Preview: Command + P
// Learn more: https://framer.com/api

export function Key(props) {
    // const onTap = value => {
    //     console.log(value)
    // }

    return (
        <Frame
            background={props.tint}
            whileTap={{
                scale: 1.3,
                zIndex: 99,
            }}
            width={props.keyWidth}
            height={props.keyHeight}
            radius={4.6}
            style={{
                boxShadow: "0 1px 0 0 #898A8D",
                fontSize: props.FontSize,
                color: "#000000",
                letterSpacing: -0.55,
                textAlign: "center",
                display: "flex",
                justifyContent: "center",
                alignItems: "center",
                marginLeft: props.marginL,
                marginRight: props.marginR,
            }}
            onTap={props.onTap}
        >
            {props.content == "Shift"
                ? shiftKey
                : props.content == "Back"
                ? backKey
                : props.content}
        </Frame>
    )
}

Key.defaultProps = {
    content: "A",
    keyHeight: 42,
    keyWidth: 32,
    FontSize: 22.5,
    tint: "#FCFCFE",
}
